src/warc2zim/utils.py

```diff
@@ -42,14 +53,61 @@ def parse_title(content):
         return ""
 
 
-def to_string(input_: str | bytes) -> str:
-    try:
-        input_ = input_.decode(  # pyright: ignore[reportGeneralTypeIssues, reportAttributeAccessIssue]
-            "utf-8-sig"
-        )
-    except AttributeError:
-        pass
-    return input_  # pyright: ignore[reportGeneralTypeIssues, reportReturnType]
+def get_record_encoding(record: ArcWarcRecord) -> str | None:
+    content_type = get_record_content_type(record)
+    if m := ENCODING_RE.search(content_type):
+        return m.group("encoding")
+
+
+def to_string(input_: str | bytes, encoding: str | None) -> str:
+    """
+    Decode content to string, trying to be the more tolerant possible to invalid
+    declared encoding.
+
+    This try decode the content using 3 methods:
+     - From http headers in the warc record (given as `encoding` argument)
+     - From encoding declaration inside the content (hopping that content can be
+       losely decode using ascii to something usable)
+     - From statistical analysis of the content (made by chardet)
+
+    """
+    tried_encodings = set()
+    if isinstance(input_, str):
+        return input_
+
+    if not input_:
+        # Empty bytes are easy to decode
+        return ""
+
+    if encoding:
+        try:
+            return input_.decode(encoding)
+        except (ValueError, LookupError):
+            tried_encodings.add(encoding)
+            pass
+
+    # Detect encoding from content.
+    content_start = input_[:1024].decode("ascii", errors="replace")
+    if m := ENCODING_RE.search(content_start):
+        encoding = m.group("encoding")
+        if encoding and encoding not in tried_encodings:
```

Comment:
I haven't put a pragma here because it is not something that we shouldn't test.
We should but it kinda needs some complex test and we don't really care. But it is a real feature with a code path production code can takes.

It is not the same as test for debugging purpose or really rare conditions.

We are not testing 100% of the patch, that's a reality. Let's not mask it.
(But I will not argue more, if you want it, you will have it)

Reply:
I'm fine with it (or without it more exactly) 🤣